exercise/submission_models.py

```diff
@@ -28,18 +29,29 @@ def get_queryset(self):
             .prefetch_related('submitters')
 
     def create_from_post(self, exercise, submitters, request):
+
+        submission_data_list = [
+                (key, value) for (key, value) in query_dict_to_list_of_tuples(request.POST)
+                if key != '__aplus__'
+        ]
+        try:
+            meta_data_dict = json.loads(request.POST.get('__aplus__','{}'))
+        except json.JSONDecodeError:
+            raise ValueError("invalid json for __aplus__")
+
         new_submission = Submission.objects.create(
             exercise=exercise,
-            submission_data=query_dict_to_list_of_tuples(request.POST)
+            submission_data=submission_data_list,
+            meta_data=meta_data_dict,
         )
         new_submission.submitters.set(submitters)
         try:
             new_submission.add_files(request.FILES)
-        except DatabaseError:
+        except DatabaseError as error:
             logger.exception("Failed to save submitted files: %s %s",
                 request.user.username, exercise);
             new_submission.delete()
-            return None
+            raise DatabaseError from error
```

Comment:
Interesting. I have been wrong. The `from error` is not actually must in python 3, however it should be prefered due to semantic difference (also recommended in PEP 8). See the different message in following examples (during vs. direct cause):

```python
>>> try:
...   raise ValueError('foo')
... except ValueError:
...   raise RuntimeError()
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ValueError: foo

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
RuntimeError
```

compared to 

```python
>>> try:
...   raise ValueError('foo')
... except ValueError as error:
...   raise RuntimeError from error
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ValueError: foo

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
RuntimeError
```

To remove the chaining information would require syntax:

```python
try:
  ...
except Exception:
  raise AnyException from None
```

To conclude, this line is now ok.

Reply:
Is it OK? At the moment, I see `raise DatabaseError from error` and doesn't that instantiate a new exception even though the same exception could be re-raised?

Reply:
This should be good now